packages/jupyter-ai/src/inline-completions.ts

```diff
@@ -0,0 +1,348 @@
+      const activeCell = context.widget.content.activeCell;
+      if (activeCell) {
+        cellId = activeCell.model.id;
+      }
+    }
+    if (!path && context.widget instanceof DocumentWidget) {
+      path = context.widget.context.path;
+    }
+    const number = ++this._counter;
+    const language = this.options.languageRegistry.findByMIME(mime);
+    const result = await this.options.completionHandler.sendMessage({
+      path: context.session?.path,
+      mime,
+      prefix: this._prefixFromRequest(request),
+      suffix: this._suffixFromRequest(request),
+      language: this._resolveLanguage(language),
+      number,
+      cell_id: cellId
+    });
+    const error = result.error;
+    if (error) {
+      Notification.emit(`Inline completion failed: ${error.type}`, 'error', {
+        autoClose: false,
+        actions: [
+          {
+            label: 'Show Traceback',
+            callback: () => {
+              showErrorMessage(
+                'Inline completion failed on the server side',
+                error.traceback
+              );
+            }
+          }
+        ]
+      });
+      throw new Error(
+        `Inline completion failed: ${error.type}\n${error.traceback}`
+      );
+    }
+    return result.list;
+  }
+
+  get schema(): ISettingRegistry.IProperty {
+    return {
+      properties: {
+        maxPrefix: {
+          title: 'Maximum prefix length',
+          minimum: 1,
+          type: 'number',
+          description:
+            'At most how many prefix characters should be provided to the model.'
+        },
+        maxSuffix: {
+          title: 'Maximum suffix length',
+          minimum: 0,
+          type: 'number',
+          description:
+            'At most how many suffix characters should be provided to the model.'
+        }
+      },
+      default: JupyterAIInlineProvider.DEFAULT_SETTINGS as any
+    };
+  }
+
+  async configure(settings: { [property: string]: JSONValue }): Promise<void> {
+    this._settings = settings as unknown as JupyterAIInlineProvider.ISettings;
+  }
+
+  /**
+   * Extract prefix from request, accounting for context window limit.
+   */
+  private _prefixFromRequest(request: CompletionHandler.IRequest): string {
+    const textBefore = request.text.slice(0, request.offset);
+    const prefix = textBefore.slice(
+      -Math.min(this._settings.maxPrefix, textBefore.length)
+    );
+    return prefix;
+  }
+
+  /**
+   * Extract suffix from request, accounting for context window limit.
+   */
+  private _suffixFromRequest(request: CompletionHandler.IRequest): string {
+    const textAfter = request.text.slice(request.offset);
+    const prefix = textAfter.slice(
+      0,
+      Math.min(this._settings.maxPrefix, textAfter.length)
+    );
+    return prefix;
+  }
+
+  private _resolveLanguage(language: IEditorLanguage | null) {
+    if (!language) {
+      return 'plain English';
+    }
+    if (language.name === 'ipythongfm') {
```

Comment:
We could add `ipython` → `python` mapping to make sure weaker models understand but then we are loosing some specificity. Maybe best would be making it a user-configurable mapping to allow users to tune it themselves (here or in a follow-up PR). What do you think?

Reply:
Seems like it's more reasonable to prefer `python` over `ipython` by default. While it does lose specificity, I think that it's pretty easy for models to get confused between the two because of how similar they are in syntax.

I also think that there's no need to make this configurable in this PR. My preference is to add configurability after the core feature is made available. Otherwise, we could be adding complexity to no user benefit.

Reply:
Thank you!

```suggestion
    if (language.name === 'ipython') {
      return 'python';
    }
    else if (language.name === 'ipythongfm') {
```